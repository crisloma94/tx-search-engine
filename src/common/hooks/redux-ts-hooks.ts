import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';
import { AppDispatch, RootState } from '../../core/store/store';

// Necessary for correct ts store types
export const useAppDispatch = () => useDispatch<AppDispatch>();
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
