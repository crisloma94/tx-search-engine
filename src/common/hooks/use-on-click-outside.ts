import { useEffect, Ref } from 'react';

export const useOnClickOutside = (
  ref: Ref<HTMLDivElement>,
  handler: (event: MouseEvent | TouchEvent) => void,
) => {
  useEffect(() => {
    const listener = (event: MouseEvent | TouchEvent) => {
      if (
        ref &&
        'current' in ref &&
        (!ref.current || ref.current.contains(event.target as Node))
      ) {
        return;
      }

      handler(event);
    };

    document.addEventListener('mousedown', listener);
    document.addEventListener('touchstart', listener);

    return () => {
      document.removeEventListener('mousedown', listener);
      document.removeEventListener('touchstart', listener);
    };
  }, [ref, handler]);
};
