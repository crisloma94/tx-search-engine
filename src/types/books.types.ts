// Defined in DEFAULT_BOOK_FIELDS
export interface BookSearchModel {
  author_name?: string[];
  cover_i?: number;
  first_publish_year?: number;
  id_goodreads?: string[];
  id_librarything?: string[];
  isbn?: string[];
  key: string;
  oclc?: string[];
  publisher?: string[];
  title: string;
}
