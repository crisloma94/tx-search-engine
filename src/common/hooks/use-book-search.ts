import { useEffect, useState } from 'react';
import { BookSearchModel } from '../../types/books.types';
import axios from 'axios';
import { BOOK_SEARCH_API_URL } from '../constants/api';
import { DEFAULT_BOOK_FIELDS, DEFAULT_LIMIT_PARAM } from '../constants/request';

interface useBookSearchProps {
  query: string;
}

export const useBookSearch = ({ query }: useBookSearchProps) => {
  const [offsetParam, setOffsetParam] = useState<number>(0);
  const [totalResults, setTotalResults] = useState<number>(0);
  const [results, setResults] = useState<BookSearchModel[]>([]);
  const [loading, setLoading] = useState<boolean>(false);
  const [error, setError] = useState<string | null>(null);

  useEffect(() => {
    const fetchData = async () => {
      setLoading(true);
      setError(null);
      try {
        if (query.length > 2) {
          const response = await axios.get(`${BOOK_SEARCH_API_URL}`, {
            params: {
              q: query,
              fields: DEFAULT_BOOK_FIELDS,
              limit: DEFAULT_LIMIT_PARAM,
              offset: offsetParam,
            },
          });
          setResults((previousResults) => [
            ...previousResults,
            ...response.data.docs,
          ]);
          setTotalResults(response.data.numFound);
          setLoading(false);
        }
      } catch (error) {
        setError('There has been an error searching for books');
        setLoading(false);
      }
    };

    fetchData();
  }, [offsetParam, query]);

  const loadMore = () => {
    const newOffset = offsetParam + DEFAULT_LIMIT_PARAM;
    if (newOffset <= totalResults) {
      setOffsetParam(newOffset);
    }
  };

  const clearSearch = () => {
    setResults([]);
    setTotalResults(0);
    setOffsetParam(0);
    setLoading(false);
    setError(null);
  };

  return { results, loading, error, loadMore, clearSearch };
};
