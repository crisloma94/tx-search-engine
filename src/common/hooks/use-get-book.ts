import { useEffect, useState } from 'react';
import { BookSearchModel } from '../../types/books.types';
import axios from 'axios';
import { BOOK_SEARCH_API_URL } from '../constants/api';
import { DEFAULT_BOOK_FIELDS } from '../constants/request';

export const useGetBook = (key: string) => {
  const [bookData, setBookData] = useState<BookSearchModel>(
    {} as BookSearchModel,
  );
  const [loading, setLoading] = useState<boolean>(false);
  const [error, setError] = useState<string | null>(null);

  useEffect(() => {
    const fetchData = async () => {
      setLoading(true);
      setError(null);
      try {
        const response = await axios.get(`${BOOK_SEARCH_API_URL}`, {
          params: { q: key, fiedls: DEFAULT_BOOK_FIELDS, limit: 1, offset: 0 },
        });
        setBookData(response.data.docs[0]);
      } catch (error) {
        setError('An error has ocurred fetching book detail data');
      }

      setLoading(false);
    };

    fetchData();
  }, [key]);

  return { bookData, loading, error };
};
