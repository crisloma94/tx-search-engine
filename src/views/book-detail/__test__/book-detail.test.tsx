import { render, screen } from '@testing-library/react';
import Theme from '../../../styles/theme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import BookDetail from '../book-detail.view';
import * as hook from '../../../common/hooks/use-get-book';
import { BookSearchModel } from '../../../types/books.types';

const defaultBook = {
  author_name: ['Author name test'],
  cover_i: 5678,
  first_publish_year: 1984,
  id_goodreads: ['1234'],
  id_librarything: ['567'],
  isbn: ['89'],
  key: '/works/OL545572W',
  oclc: ['1', '2'],
  publisher: ['Publisher example'],
  title: 'Title test',
};

const mockStore = configureStore([]);
const mockedStore = mockStore({});

describe('Book detail view test suite', () => {
  const getComponent = () => (
    <Provider store={mockedStore}>
      <Theme>
        <BookDetail />
      </Theme>
    </Provider>
  );
  const setup = () => render(getComponent());

  it('Should match snapshot', () => {
    jest.spyOn(hook, 'useGetBook').mockImplementation(() => ({
      bookData: defaultBook,
      error: '',
      loading: false,
    }));
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });

  it('Should add a book on page load', () => {
    jest.spyOn(hook, 'useGetBook').mockImplementation(() => ({
      bookData: defaultBook,
      error: '',
      loading: false,
    }));
    setup();
    expect(screen.getByText('Author name test')).toBeInTheDocument();
  });

  it('Should render an error message when there is an error fetching the data', () => {
    jest.spyOn(hook, 'useGetBook').mockImplementation(() => ({
      bookData: {} as BookSearchModel,
      error: 'Error',
      loading: false,
    }));
    setup();
    expect(
      screen.getByText('Sorry... there has been an error fetching the data'),
    ).toBeInTheDocument();
  });
});
