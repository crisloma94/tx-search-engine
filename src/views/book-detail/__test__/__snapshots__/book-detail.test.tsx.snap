// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Book detail view test suite Should match snapshot 1`] = `
<div>
  <div
    class="sc-gTRrQi fUVnky"
  >
    <div
      class="sc-ewnqHT ZpWIw"
    >
      <img
        alt="Default book cover"
        class="sc-bgqQcB ccLtkH book-cover"
        src="https://covers.openlibrary.org/b/id/5678-M.jpg"
      />
      <div
        class="book-detail__info"
      >
        <h1
          class="book-detail__title"
        >
          Title test (1984)
        </h1>
        <h2
          class="book-detail__author"
        >
          Author name test
        </h2>
        <p
          class="book-detail__publisher"
        >
          Publisher example
        </p>
        <a
          href="https://www.amazon.es/s?k=Title test+Author name test"
          target="_blank"
        >
          Search on amazon
        </a>
      </div>
    </div>
  </div>
</div>
`;
